manual/manual/html_processing/src/process_api.ml

```diff
@@ -0,0 +1,414 @@
+        |> prepend_child body;
+    | None ->
+        if search then search_widget false |> prepend_child nav;
+        (* Add "general index" link to all other files *)
+        create_element "a" ~inner_text:"< General Index"
+          ~attributes:["href", "index.html"]
+        |> prepend_child nav in
+
+  (* Add version number *)
+  add_version_link nav ((if !compiler_libref then "Compiler " else "") ^
+                        "API Version " ^ version) releases_url;
+
+  (* Add logo *)
+  prepend_child header (logo_html
+                          ((if !compiler_libref then "../" else "") ^
+                           (manual_page_url ^ "/index.html")));
+
+  sprintf "Saving %s..." out |> pr;
+
+  (* Save new html file *)
+  let new_html = to_string soup in
+  write_file out new_html
+
+let process ?(overwrite=false) ~version file out =
+  if overwrite || not (Sys.file_exists out)
+  then Ok (process ~version file out)
+  else Error (sprintf "File %s already exists." out)
+
+let all_html_files () =
+  Sys.readdir !src_dir |> Array.to_list
+  |> List.filter (fun s -> Filename.extension s = ".html")
+
+
+(* Creation of a search bar *)
+let parse_pair = function
+  | [a; b] -> (a,b)
+  | _ -> raise (Invalid_argument "parse_pair")
+
+let parse_tdlist = function
+  | [alist, []] -> let mdule, value = parse_pair alist in
+    pr (fst value);
+    (mdule, value, ("",""))
+  | [[], infolist; alist, []] ->
+    let mdule, value = parse_pair alist in
+    let infohtml, infotext = List.split infolist in
+    let infohtml = infohtml |> List.map to_string |> String.concat " "
+                   |> String.escaped in
+    let infotext = infotext |> String.concat " " |> String.escaped in
+    pr (fst value);
+    (mdule, value, (infohtml, infotext))
+  | _ -> raise (Invalid_argument "parse_tdlist")
+
+(* Generate the index using Lambdasoup ==> very slow, see below. *)
+let make_index () =
+  let html = read_file ("index_values.html"
+                        |> with_dir !src_dir) in
+  let soup = parse html in
+  soup $ "table"
+  |> select "tr"
+  |> fold (fun trlist tr ->
+      let tdlist =
+        tr $$ "td"
+        |> fold (fun tdlist td ->
+            let alist = td $$ ">a"
+                        |> fold (fun alist a ->
+                            (R.leaf_text a, R.attribute "href" a) :: alist
+                          ) [] in
+            let infolist = td $$ "div.info"
+                           |> to_list
+                           |> List.map (fun info ->
+                               let infotext = texts info
+                                              |> String.concat "" in
+                               (info, infotext)) in
+            if alist = [] && infolist = [] then tdlist
+            else (alist, infolist) :: tdlist
+          ) [] in
+      if tdlist = [] then trlist
+      else (parse_tdlist tdlist) :: trlist
+    ) []
+
+
+(* Fast version of make_index using Scanf *)
```

Comment:
You don't seem to use either version? What difference in speed are we talking here?

Reply:
Scanf version is about 20x faster

Reply:
following your suggestion, I finally opted for the slow version, but with a cache to speed up a little bit (on my laptop, the slow version takes 18 sec with caching, 50 sec without, and the fast version takes 1 sec).